Sources/StreamChat/Repositories/ConnectionRepository.swift

```diff
@@ -0,0 +1,225 @@
+//
+// Copyright © 2022 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+class ConnectionRepository {
+    private let connectionQueue: DispatchQueue = DispatchQueue(label: "io.getstream.connection-repository", attributes: .concurrent)
+    private var _connectionIdWaiters: [String: (String?) -> Void] = [:]
+    private var _connectionId: ConnectionId?
+    private var _connectionStatus: ConnectionStatus = .initialized
+
+    private var connectionIdWaiters: [String: (String?) -> Void] {
+        get { connectionQueue.sync { _connectionIdWaiters } }
```

Comment:
Not in the scope of this PR, but food for thought: since we will be adding more thread-safe repos in the future, maybe we should think about abstracting this logic for getters and setters, since it will repeat itself?

Reply:
I did not do that because it requires us to make that wrapping class as public in some cases...
But might be trying to find a workaround on that soon

Reply:
I think writing Repositories as concrete classes usually are not the best practice. Repositories by definition are interface adapters, they abstract how we fetch data. By creating concrete classes instead it does not give us the flexibility that repositories should give. With interfaces, we can easily create Decorators, which abstract the Threading code. This will also improve a lot how we test this. 

As abstracting all together thread-safe code, I don't think it will be possible, unless we go back to @Atomic 🤔 Which we want to move away from.